social_engagement/handlers.py

```diff
@@ -11,107 +9,146 @@
     thread_deleted,
     comment_deleted,
     thread_voted,
-    comment_voted,
     thread_followed,
     thread_unfollowed,
+    thread_or_comment_flagged,
 )
 import lms.lib.comment_client as cc
-from social_engagement.tasks import task_update_user_engagement_score
+from social_engagement.tasks import task_handle_change_after_signal
 
 
 @receiver(thread_deleted)
-@receiver(comment_deleted)
 @receiver(thread_voted)
-@receiver(comment_voted)
-def thread_comment_signal_handler(sender, **kwargs):  # pylint: disable=unused-argument
+def thread_signal_handler(sender, **kwargs):  # pylint: disable=unused-argument
     """
-    Updates user social engagement score
+    Updates user social engagement score for deleted and voted thread (or voted comment).
+    TODO: And also, for some reason, on voted comment.
     """
-    post = kwargs['post']
-    course_id = getattr(post, 'course_id', None)
-    user_id = getattr(post, 'user_id', None)
-    social_engagement_enabled = settings.FEATURES.get('ENABLE_SOCIAL_ENGAGEMENT', False)
-
-    if social_engagement_enabled:
-        if course_id and user_id:
-            task_update_user_engagement_score.delay(unicode(course_id), user_id)
+    thread = kwargs['post']
+    course_id = getattr(thread, 'course_id', None)
+    user_id = getattr(thread, 'user_id', None)
 
-        # present only in two cases: thread_deleted & comment_deleted
-        if 'involved_users' in kwargs:
-            involved_users = kwargs['involved_users']
-            users = User.objects.filter(username__in=involved_users).exclude(id=user_id)
+    # present if thread_deleted
+    if 'involved_users' in kwargs:
+        users = kwargs['involved_users']
+        for user, user_data in users.items():
+            _decrement(user, course_id, user_data)
 
-            for user in users:
-                task_update_user_engagement_score.delay(unicode(course_id), user.id)
+    # thread or comment voted
+    else:
+        change = _decrement if kwargs.get('undo') else _increment
+        change(user_id, course_id, 'num_upvotes')
 
 
 @receiver(thread_created)
 def thread_created_signal_handler(sender, **kwargs):  # pylint: disable=unused-argument
     """
-    Updates user social engagement score
+    Updates user social engagement score for created thread.
     """
     thread = kwargs['post']
     course_id = getattr(thread, 'course_id', None)
     action_user = kwargs['user']
-    social_engagement_enabled = settings.FEATURES.get('ENABLE_SOCIAL_ENGAGEMENT', False)
```

Comment:
Similar comment here and elsewhere: we probably shouldn't activate the job if the feature is disabled. I understand that it decreases code repetition to put it in the jobs rather than everywhere else, but it makes the performance worse for celery and the queue because they may get a bunch of unnecessary jobs while the feature is disabled.